connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,220 @@
+mutistream-selection negotiation for muxer selection.
+
+In order to achieve the above stated goal, each candidate muxer will be
+represented by a protocol name/code, and the candidate muxers are supplied to
+the security protocol's handshake process as a list of protocol names.
+
+If the client and server agree upon the common muxer to be used, then the
+result of the muxer selection is a muxer code represented by the selected
+protocol name. If no agreement is reached upon by the client and server
+then an empty muxer code is returned and the connection upgrade process
+MUST fall back to the multistream-selection protocol to negotiate the muxer.
+
+
+### Muxer selection over TLS
+
+When the security protocol selected by the upgrader is TLS, the [ALPN]
+extesion of TLS handshake is used to select the muxer.
+
+   With ALPN, the client sends the list of supported application
+   protocols as part of the TLS ClientHello message.  The server chooses
+   a protocol and sends the selected protocol as part of the TLS
+   ServerHello message.  The application protocol negotiation can thus
+   be accomplished within the TLS handshake, without adding network
+   round-trips, and allows the server to associate a different
+   certificate with each application protocol, if desired.
+
+For the purpose of muxer selection, the types of muxers are coded as protocol
+names in the form of a list of strings, and inserted in the ALPN "NextProtos"
+field. An example list as following:
+
+    ["yamux/1.0.0", "mplux", "libp2p"]
+
+The NextProtos list is ordered by preference, with the most prefered muxer at
+the beginning. The "libp2p" protocol code MUST always be the last item in the
+ALPN NextProtos list.
+
+The server chooses the supported protocol by going through its prefered
+protocol list and searchs if the protocol is supported by the client too. if no
+mutually supported protcol is found the TLS handshake will fail.
+
+If the selected NextProto is "libp2p" then the muxer selection process returns
+an empty result, and the multistream-selection protocol MUST be run to negotiate
+the muxer.
+
+(TBD: in some cases early abortion is desireable)
+
+
+### Muxer selection over Noise
+
+The libp2p Noise implementation allows the Noise handshake process to carry
+early data. [Noise-Early-Data] is carried in the second and third message of
+the handshake pattern:
+
+    XX:
+    -> e
+    <- e, ee, s, es
+    -> s, se
+
+At the end of the handshake pattern, both the client and server have received
+the peer's early data. The Noise protocol does not perform the protocol
+selection as TLS does, rather it just delivers the early data to handshaking
+peers.
+
+The muxer selection logic runs out of the Noise handshake process, relying on
+the early data exchanged during the handshake. The early data is delivered in
+the form of a byte string. The supported muxers are passed in space separated
+string codes. An example early data string:
```

Comment:
I am very much in favor of using a protobuf here for the above mentioned reasons.

How about taking this one step further and use the _Protocol Select_ (https://github.com/libp2p/specs/pull/349) protobuf definition? To me this proposal actually reads as an extension to _Protocol Select_. Would people agree with that?

Reply:
I’m not convinced that would be a good idea.

1. This is specific to just the Noise handshake, there’s no reason to cause any coupling here
2. The Protocol Select spec might change before we actually deploy it, which would leave us with a zombie protobuf here

Reply:
> This is specific to just the Noise handshake

Related to this discussion, would it not make sense to use the same structure for both Noise and TLS? In case TLS only allows valid UTF-8 strings, we can use `multibase`.

Reply:
> The Protocol Select spec might change before we actually deploy it, which would leave us with a zombie protobuf here

In case it changes, we can do those changes in a backwards compatible way, given that we are using Protobuf, no?

Reply:
I am still failing to see the difference between the Protocol Select protocol and this proposal. The only difference to me is that this proposal exchanges the messages via the early data mechanism, whereas the Protocol Select protocol exchanges the messages on an _established_ channel. Am I missing something?

Both protocols
- Exchange a list of protocols
- Use the same prioritization algorithm
- Fail in the same scenarios (no protocols overlap)

How about adding an optional early-negotiation mechanism to Protocol Select, namely the exchange described in this pull request.

Reply:
> Related to this discussion, would it not make sense to use the same structure for both Noise and TLS? In case TLS only allows valid UTF-8 strings, we can use multibase.

That doesn't work with existing libraries, who do ALPN negotiation for you. They expect a certain format.

> In case it changes, we can do those changes in a backwards compatible way, given that we are using Protobuf, no?

I still don't see the advantage of using a more complicated than necessary Protobuf here. This negotiation step is very tightly bound to the handshake protocol, and it's reasonable to expect that different handshake protocols will use different encodings (as TLS and Noise already do).

> The only difference to me is that this proposal exchanges the messages via the early data mechanism, whereas the Protocol Select protocol exchanges the messages on an established channel. Am I missing something?

That's a huge difference. It saves 1-RTT *in every case*. In specific, early data in TLS 1.3 only works on resumed connections, whereas this proposal works on every connection.

Reply:
I like this section of discussion, it clarifies lot of things. 
I will add a protobuf definition for Noise.

Reply:
> > The only difference to me is that this proposal exchanges the messages via the early data mechanism, whereas the Protocol Select protocol exchanges the messages on an established channel. Am I missing something?
> 
> That's a huge difference. It saves 1-RTT _in every case_. In specific, early data in TLS 1.3 only works on resumed connections, whereas this proposal works on every connection.

Sorry. I think I phrased this in a confusing way. I am definitely in favor of the 1RTT saving. My point was that I don't see the difference in the two that would justify different encodings. With the restriction on TLS' ALPN mechanism, this makes sense now.

> > Related to this discussion, would it not make sense to use the same structure for both Noise and TLS? In case TLS only allows valid UTF-8 strings, we can use multibase.
> 
> That doesn't work with existing libraries, who do ALPN negotiation for you. They expect a certain format.

Thanks for raising this. I was not aware of (/ didn't think of) this restriction. It is the case for Rust, I am guessing it is also the case for the other languages and their major TLS implementations.

Let's use the current proposal and not the Protocol Select Protobuf.